subxt/src/metadata/decode_encode_traits.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright 2019-2022 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+use super::Metadata;
+use crate::error::Error;
+
+/// This trait is implemented for all types that also implement [`scale_decode::DecodeAsType`].
+pub trait DecodeWithMetadata: Sized {
+    /// Given some metadata and a type ID, attempt to SCALE decode the provided bytes into `Self`.
+    fn decode_with_metadata(
+        bytes: &mut &[u8],
+        type_id: u32,
+        metadata: &Metadata,
+    ) -> Result<Self, Error>;
+}
+
+impl<T: scale_decode::DecodeAsType> DecodeWithMetadata for T {
+    fn decode_with_metadata(
+        bytes: &mut &[u8],
+        type_id: u32,
+        metadata: &Metadata,
+    ) -> Result<T, Error> {
+        let val = T::decode_as_type(bytes, type_id, metadata.types())?;
+        Ok(val)
+    }
+}
+
+/// This trait is implemented for all types that also implement [`scale_encode::EncodeAsType`].
+pub trait EncodeWithMetadata {
+    /// SCALE encode this type to bytes, possibly with the help of metadata.
+    fn encode_with_metadata(
+        &self,
+        type_id: u32,
+        metadata: &Metadata,
+        bytes: &mut Vec<u8>,
+    ) -> Result<(), Error>;
+}
+
+impl<T: scale_encode::EncodeAsType> EncodeWithMetadata for T {
+    /// SCALE encode this type to bytes, possibly with the help of metadata.
+    fn encode_with_metadata(
+        &self,
+        type_id: u32,
+        metadata: &Metadata,
+        bytes: &mut Vec<u8>,
+    ) -> Result<(), Error> {
+        self.encode_as_type_to(type_id, metadata.types(), bytes)?;
```

Comment:
I couldn't remember until I tried removing them, but one type in particular (`DecodedValueThunk`, which is just a thing that hasn't been decoded yet but can be turned into a dynamic value or decoded manually into some static type) can implement `DecodeWithMetadata` but not `DecodeAsType` (since it wants  to store the `Metadata`, but `DecodeAsType` provides the lower level `PortableRegistry`). I remember trying to work around this when I ran into it, but it just made most sense to have the higher level trait in the end.